src/main/java/com/bloxbean/cardano/client/coinselection/impl/DefaultUtxoSelectionStrategyImpl.java

```diff
@@ -1,116 +1,149 @@
-    public List<Utxo> selectUtxos(String address, String unit, BigInteger amount, Set<Utxo> excludeUtxos) throws ApiException {
-        return selectUtxos(address, unit, amount, null, excludeUtxos);
+    public DefaultUtxoSelectionStrategyImpl(UtxoService utxoService, boolean ignoreUtxosWithDatumHash) {
+        this.utxoService = utxoService;
+        this.ignoreUtxosWithDatumHash = ignoreUtxosWithDatumHash;
     }
 
     @Override
-    public List<Utxo> selectUtxos(String address, String unit, BigInteger amount, String datumHash, Set<Utxo> excludeUtxos) throws ApiException {
-        if(amount == null)
-            amount = BigInteger.ZERO;
-
-        BigInteger totalUtxoAmount = BigInteger.valueOf(0);
-        List<Utxo> selectedUtxos = new ArrayList<>();
-        boolean canContinue = true;
-        int i = 1;
-
-        while(canContinue) {
-            Result<List<Utxo>> result = utxoService.getUtxos(address, getUtxoFetchSize(),
-                    i++, getUtxoFetchOrder());
-            if(result.code() == 200) {
-                List<Utxo> fetchData = result.getValue();
-
-                List<Utxo> data = filter(fetchData);
-                if(data == null || data.isEmpty()) {
-                    canContinue = false; //Result code 200, but no utxo returned
-                    throw new ApiException(String.format("Unable to get enough Utxos for address : %s, reason: %s", address, result.getResponse()));
-                }
-
-                for(Utxo utxo: data) {
-                    if(excludeUtxos.contains(utxo))
+    public Set<Utxo> select(String sender, List<Amount> outputAmounts, String datumHash, Set<Utxo> utxosToExclude, int maxUtxoSelectionLimit) {
+        if(outputAmounts == null || outputAmounts.isEmpty()){
+            return Collections.emptySet();
+        }
+        try{
+            // loop over utxo's, find matching requested amount
+            Set<Utxo> selectedUtxos = new HashSet<>();
+
+            final Map<String, BigInteger> remaining = new HashMap<>(outputAmounts.stream()
+                    .collect(Collectors.groupingBy(Amount::getUnit,
+                            Collectors.reducing(BigInteger.ZERO,
+                                    Amount::getQuantity,
+                                    BigInteger::add))))
+                    .entrySet().stream()
+                    .filter(entry -> BigInteger.ZERO.compareTo(entry.getValue()) < 0)
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            int page = 0;
+            final int nrOfItems = 100;
+
+            while(!remaining.isEmpty()){
+                var fetchResult = utxoService.getUtxos(sender, nrOfItems, page + 1, OrderEnum.asc);
+
+                var fetched = fetchResult != null && fetchResult.getValue() != null
+                        ? fetchResult.getValue()
+                                     .stream()
+                                     .sorted(sortByMostMatchingAssets(outputAmounts))
```

Comment:
Should we start calling this default impl as "MostMatchedAssetSequentialStrategy" :)

Reply:
The reason I added `sort` was to mimic the existing logic. 

In the current `UtxoTransactionBuilderImpl` on master, when processing a multi-asset `PaymentTransaction`, first the utxos are selected for the native token and only after that for min lovelace.

But for this PR, all assets are send to the selection strategy together and utxos are only fetched once.
But processing them in natural order could result in more tx's being selected. For example, if there is a UTXO with enough native tokens and ada to cover all requested amounts but another UTXO exist (before this one) with only ADA, then that one will be picked for covering the min ADA and the other one for covering the native token.

in order to achieve the same result as before, this sort was added, but I'll leave it up to you to rename (or change) as you wish :-)